### ISSUE TITLE
ncclient get_config timing out

### ISSUE THREAD

**Issue by allan-silverstein** [ncclient/ncclient#436]
Hi,
I'm trying to do a simple ncclient get_config but I'm getting an ncclient timeout error.  The vendor is from Drivenets.  I'm able to exchange capabilities and I'm able to pull the config when I do a manual get-config with XML.  I've tried increasing the timeout to 120 seconds but same result.  I also have a tcpdump running on the Linux server interface and when I issue the get_config method I see a bunch of packets coming back from the device (similar packets to when I perform the get-config manually with XML).  

Here is the error:
```
traceback (most recent call last):
  File "/home/as2863/pythonProjects/nettools/netconf/getconfig.py", line 17, in <module>
    running_config = m.get_config('running')
  File "/home/as2863/python-venv/yang/lib/python3.7/site-packages/ncclient/manager.py", line 231, in execute
    huge_tree=self._huge_tree).request(*args, **kwds)
  File "/home/as2863/python-venv/yang/lib/python3.7/site-packages/ncclient/operations/retrieve.py", line 166, in request
    return self._request(node)
  File "/home/as2863/python-venv/yang/lib/python3.7/site-packages/ncclient/operations/rpc.py", line 366, in _request
    raise TimeoutExpiredError('ncclient timed out while waiting for an rpc reply.')
ncclient.operations.errors.TimeoutExpiredError: ncclient timed out while waiting for an rpc reply.
```

Here is the simple script:

```
from ncclient import manager
from pprint import pprint
import xmltodict
import xml.dom.minidom

router = {
   'host': 'rtr1',
   'port': '830',
   'username': 'xxx',
   'password': 'xxx'
}

m = manager.connect(host=router['host'], port=router['port'], username=router['username'],
                    password=router['password'], hostkey_verify=False)

running_config = m.get_config('running')
print(xml.dom.minidom.parseString(running_config).toprettyxml())

m.close_session()
```

Any suggestion for how to troubleshoot this?  I'm almost positive the device is responding with the config because I also have a debug running on the device and see the XML message that is "supposedly" being sent to ncclient.  The only thing that bothers me a little is the MTU on the Linux server is set at 1500 but some of the packets being returned are over 7000 bytes.  Don't ask... This is lab environment.  Regardless,  The XML formatted config is returned almost instantly when getting the config interactively (not using ncclient) using the following:

```
(yang) al@ubuntuP1-SYSlog-S1:~/pythonProjects/nettools/netconf$ ssh al@rtr1 -p830

al@rtr1's password:
<?xml version="1.0"?><hello xmlns="urn:ietf:params:xml:ns:netconf:base:1.0"><capabilities><capability>urn:ietf:params:netconf:base:1.0</capability><capability>urn:ietf:params:netconf:capability:candidate:1.0</capability><capability>urn:ietf:params:netconf:capability:validate:1.0</capability><capability>urn:ietf:params:netconf:capability:confirmed-commit:1.0</capability><capability>urn:ietf:params:netconf:base:1.1</capability><capability>urn:ietf:params:netconf:capability:with-defaults:1.0?basic-mode=explicit&amp;also-supported=report-all</capability></capabilities><session-id>1</session-id></hello>]]>]]>
<?xml version="1.0" encoding="UTF-8"?>
<hello xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
    <capabilities>
        <capability>urn:ietf:params:netconf:base:1.0</capability>
        <capability>urn:ietf:params:netconf:base:1.1</capability>
        <capability>urn:ietf:params:netconf:capability:xpath:1.0</capability>
        <capability>urn:ietf:params:xml:ns:yang:ietf-system</capability>
        <capability>urn:ietf:params:netconf:capability:candidate:1.0</capability>
    </capabilities>
</hello>
]]>]]>
<rpc message-id="1" xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
    <get-config>
        <source>
            <running/>
        </source>
    </get-config>
</rpc>
]]>]]>
...
(Full Configuration returned to terminal display in XML)
```

Thx
Al

**Comment by fredgan** [comment]
Hi @allan-silverstein ,
I use the sample script you attached to run in Netopeer2, and it will report some errors.
```
Traceback (most recent call last):
  File "D:/workspace/ncclient/issue/issue_436.py", line 19, in <module>
    print(xml.dom.minidom.parseString(running_config).toprettyxml())
  File "C:\python37\lib\xml\dom\minidom.py", line 1968, in parseString
    return expatbuilder.parseString(string)
  File "C:\python37\lib\xml\dom\expatbuilder.py", line 925, in parseString
    return builder.parseString(string)
  File "C:\python37\lib\xml\dom\expatbuilder.py", line 223, in parseString
    parser.Parse(string, True)
TypeError: a bytes-like object is required, not 'GetReply'
```
However, I change `running_config = m.get_config('running')` to `running_config = m.get_config('running').data_xml` and it's OK.

It seems that your Drivenets does not return rpc-reply data. As a result, the ncclient reports a timeout error. So I think that this isn't an ncclient issue. :-)

**Comment by einarnn** [comment]
@allan-silverstein , can you provide the full return data from the `ssh` sample you gave above with any sensitiove data obfuscated? That may help diagnose potential issues.

Also, I see the netconf server returns the capability `<capability>urn:ietf:params:netconf:base:1.1</capability>`? This means that `ncclient` will, by RFC standards, use the netconf 1.1 chunked encoding, but your example above use netconf 1.0 encoding (which should also work), so there may be differences there.

Finally, you might want to try `ncc` (`pip install ncc`) and perform a config get with the `--verbose` flag. That will show you the detailed interactions over the wire (you can also plagiarize the `ncc` code for the modules to enable debug logging for if you'd rather integrate the logging with your own code.

**Comment by allan-silverstein** [comment]
Sorry for not getting back sooner.  This issue was resolved. The vendor stated that there was a particular bug in ncclient versions 0.6.6 and below that affected them.  As a result, they modified there product to work with ncclient.  When that bug was fixed, started in 0.6.7, ncclient would no longer work with their product.  As soon as I downgraded to 0.6.6, ncclient starting working. Note, a reply was being sent back for some reason, not properly processed by ncclient (therefore it stated a timeout)
Thank you.

**Comment by einarnn** [comment]
@allan-silverstein, I would be interested in understanding the ncclient bug that impacts the vendor. Depending on what the bug is, a vendor (or anyone) can typically work around "interop" issues by providing a device- or OS-specific "plugin" for ncclient. This is used by a number of vendors to deal with things like ncclient bugs and RFC non-compliance. Rather than your vendor and you being stuck on older versions of ncclient, they could perhaps take that approach and contribute a device plugin?

Just a thought!!

**Comment by allan-silverstein** [comment]
Here is part of the email from the vendor:
Netconf 1.1 must include some special delimiters before and after each response. 
In the old ncclient versions (0.6.6 and below) that delimiter caused a bug in the client so we bypassed that by modifying these delimiters. (That's why when I loaded 0.6.6, it worked)
The bug in the ncclient was fixed in version 0.6.9 but that version will only work with the latest vendor device version which I'm not currently testing... 
I don't have any other information from the vendor...

**Comment by einarnn** [comment]
There was a bug with some Unicode character handling, but ncclient has dealt correctly with non-Unicode netconf 1.1 messages for a long time now. That sounds a bit weird, and, apart from the Unicode issue, we haven’t had any bugs reported on delimiters that I recall. Anyway, whatever it was appears to now be fixed, so that’s good.
